pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/RuntimeASTTransformer.groovy

```diff
@@ -837,7 +899,7 @@ class RuntimeASTTransformer {
             // TODO: Do I need to create a new ModelASTStage each time?  I don't think so.
             String name = "Matrix - " + cellLabels.join(", ")
 
-            return ctorX(ClassHelper.make(Stage.class),
+            return ctorXFunction(ClassHelper.make(Stage.class),
```

Comment:
I wonder how this affects variable resolution for stuff inside of the matrix, for example if someone defines a Groovy method at the top of the Jenkinsfile (not recommended, but it happens), now that the matrix stages are generated inside of a static method in a separate class, will those methods still be resolvable?

Reply:
Good question. Tried it manually and it worked. 

@abayer  
Is there a test for this scenario? (The only down side to extensive testing it not know where all the tests are.)

Reply:
There are tests for this and they caught the problem but I'm not sure that they cover all the cases.

Reply:
FWIW, I think it would be a good idea to add a system property/public static non-final variable to disable the new behavior and force the old code path in case it causes issues for existing Declarative users in scenarios that aren't covered by tests. Something like this in `RuntimeASTTransformer`:

```
@SuppressFBWarnings(warningId="the real code for the non-final static vars are bad warning", justification="For access from script console")
public static boolean FORCE_SINGLE_EXPRESSION_TRANSFORMATION = Boolean.getBoolean(RuntimeASTTransformer.class.getName() + ".FORCE_SINGLE_EXPRESSION_TRANSFORMATION");
```

And then inside of `ctorXFunction` and friends you look at `FORCE_SINGLE_EXPRESSION_TRANSFORMATION` (or whatever you want to call it) and fall back to `ctorX` if it is true.

Reply:
@dwnusbaum 
Solid idea. Done.